flashcard.koplugin/highlightparser.lua

```diff
@@ -0,0 +1,285 @@
+                    local bookmark_quote = bookmark.text:match(pattern)
+                    if bookmark_quote ~= clipping.text and bookmark.text ~= clipping.text then
+                        -- use modified quoted text or entire bookmark text if it's not a match
+                        clipping.note = bookmark_quote or bookmark.text
+                    end
+                end
+            end
+            if item.text == "" and item.pos0 and item.pos1 and item.pos0.x and item.pos0.y and item.pos1.x and
+                item.pos1.y then
+                -- highlights in reflowing mode don't have page in pos
+                if item.pos0.page == nil then
+                    item.pos0.page = page
+                end
+                if item.pos1.page == nil then
+                    item.pos1.page = page
+                end
+                local image = {}
+                image.file = book.file
+                image.pos0, image.pos1 = item.pos0, item.pos1
+                image.pboxes = item.pboxes
+                image.drawer = item.drawer
+                clipping.image = self:getImage(image)
+            end
+            --- @todo Store chapter info when exporting highlights.
+            if clipping.text and clipping.text ~= "" or clipping.image then
+                table.insert(book, {clipping})
+            end
+        end
+    end
+    table.sort(book, function(v1, v2)
+        return v1[1].page < v2[1].page
+    end)
+end
+
+function Parser:parseHistoryFile(clippings, history_file, doc_file)
+    if lfs.attributes(history_file, "mode") ~= "file" or not history_file:find(".+%.lua$") then
+        return
+    end
+    if lfs.attributes(doc_file, "mode") ~= "file" then
+        return
+    end
+    local ok, stored = pcall(dofile, history_file)
+    if ok then
+        if not stored then
+            logger.warn("An empty history file ", history_file, "has been found. The book associated is ", doc_file)
+            return
+        elseif not stored.highlight then
+            return
+        end
+        local _, docname = util.splitFilePathName(doc_file)
+        local title, author = self:getTitle(util.splitFileNameSuffix(docname), doc_file)
+        clippings[title] = {
+            file = doc_file,
+            title = title,
+            author = author
+        }
+        self:parseHighlight(stored.highlight, stored.bookmarks, clippings[title])
+    end
+end
+
+function Parser:parseHistory()
+    local clippings = {}
+    for f in lfs.dir(self.history_dir) do
+        self:parseHistoryFile(clippings, self.history_dir .. "/" .. f,
+            DocSettings:getPathFromHistory(f) .. "/" .. DocSettings:getNameFromHistory(f))
+    end
+    for _, item in ipairs(ReadHistory.hist) do
+        self:parseHistoryFile(clippings, DocSettings:getSidecarFile(item.file), item.file)
+    end
+
+    return clippings
+end
+
+function Parser:getProps(file)
+    local document = DocumentRegistry:openDocument(file)
```

Comment:
Logically this getProps function here is the only place that can be messing with the cache. Something like this may do the trick (i.e., use the *current* document if it's open so as not to cause weird interactions) but that might be against your concept with this function.

```suggestion
    local document = self.ui.document or DocumentRegistry:openDocument(file)
```